gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java

```diff
@@ -0,0 +1,322 @@
+   *     reference.
+   */
+  public IdentifiedResult collectExpiredContents(String runId) {
+    List<Row> rows = collectExpiredContentsAsDataSet(runId).collectAsList();
+    IdentifiedResult identifiedResult = new IdentifiedResult();
+    ObjectMapper objectMapper = new ObjectMapper();
+    rows.forEach(row -> addRowToIdentifiedResults(identifiedResult, objectMapper, row));
+    return identifiedResult;
+  }
+
+  /**
+   * Collect the expired contents for the given run id as spark dataset.
+   *
+   * @param runId run id of completed identify task.
+   * @return spark dataset of row where each row is having the expired contents per content id per
+   *     reference.
+   */
+  public Dataset<Row> collectExpiredContentsAsDataSet(String runId) {
+    // collect marker row for the given run-id
+    String checkMarkerRow =
+        String.format(
+            "SELECT %s FROM %s WHERE %s = '%s' AND %s = '%s'",
+            COL_GC_RUN_ID,
+            withRefName(catalogAndTable, gcRefName),
+            COL_GC_RUN_ID,
+            runId,
+            COL_TYPE,
+            TYPE_GC_MARKER);
+
+    List<Row> markerRows = sparkSession.sql(checkMarkerRow).collectAsList();
+    if (markerRows.isEmpty()) {
+      return sparkSession.createDataFrame(Collections.emptyList(), schema);
+    }
+
+    String expiredContentIdsQuery =
+        String.format(
+            "SELECT %s, %s, %s FROM %s WHERE %s = '%s' AND %s = '%s'",
+            COL_CONTENT_ID,
+            COL_EXPIRED_CONTENTS,
+            COL_EXPIRED_AT_REFERENCE_NAME,
+            //
+            withRefName(catalogAndTable, gcRefName),
+            //
+            COL_GC_RUN_ID,
+            runId,
+            //
+            COL_TYPE,
+            TYPE_CONTENT);
+
+    return sparkSession.sql(expiredContentIdsQuery);
+  }
+
+  /**
+   * Get the last written run id which is having a marker row.
+   *
+   * @return latest completed run id.
+   */
+  public String getLatestCompletedRunID() {
+    // collect marker row for the last written run id
+    String checkMarkerRow =
+        String.format(
+            "SELECT %s FROM %s WHERE %s = '%s' ORDER BY %s DESC LIMIT 1",
+            COL_GC_RUN_ID,
+            withRefName(catalogAndTable, gcRefName),
+            COL_TYPE,
+            TYPE_GC_MARKER,
+            COL_GC_RUN_START);
+    List<Row> markerRows = sparkSession.sql(checkMarkerRow).collectAsList();
+    return markerRows.isEmpty() ? null : markerRows.get(0).getString(0);
+  }
+
+  /**
+   * Convert each task result into a spark row and store it in the output table. Write a marker row
+   * in the end to indicate that the task is completed.
+   *
+   * @param resultRDD RDD of {@link IdentifiedResult}.
+   * @param runId run id of current identify task.
+   * @param startedAt task start time.
+   */
+  void writeToOutputTable(JavaRDD<IdentifiedResult> resultRDD, String runId, Timestamp startedAt) {
```

Comment:
i really dont like the RDD (sorry for hte 2nd comment) I would much rather we do this via normal spark.write tooling,